### PR TITLE
feat(hydro_deploy): Add end-to-end flamegraph generation for macos and linux localhost, fix #1351

### LINKED ISSUE #1351: feat(hydro_deploy): support profiling on macos via `dtrace`

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
Gonna sit on this for a bit and let the merge conflicts settle

File: hydro_deploy/core/src/hydroflow_crate/tracing_options.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//     },
+//     DTrace {
+
+//     },
+// }
+
+// #[cfg(test)]
+// mod test {
+
+// }
```
Comment (MingweiSamuel) [inline hydro_deploy/core/src/hydroflow_crate/tracing_options.rs:52]: remove

File: hydro_deploy/core/src/hydroflow_crate/flamegraph.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            for (i, error) in self.errors.iter().enumerate() {
+                write!(f, "({}/{}):", i + 1, self.errors.len())?;
+                error.fmt(f)?;
+                writeln!(f)?;
+            }
+            Ok(())
+        }
+    }
+}
+impl std::error::Error for MultipleErrors {}
```
Comment (MingweiSamuel) [inline hydro_deploy/core/src/hydroflow_crate/flamegraph.rs:86]: unchanged from ssh.rs

File: hydro_deploy/core/src/hydroflow_crate/perf_options.rs
Comment (MingweiSamuel) [inline hydro_deploy/core/src/hydroflow_crate/perf_options.rs:1]: renamed to TrackingOptions

File: hydroflow_plus_test/examples/perf_compute_pi.rs [hunk 3]
```diff
@@ -77,16 +78,17 @@ async fn main() {
...
                         let host = create_host(&mut deployment);
-                        let perf_options = PerfOptions::builder()
-                            .perf_outfile(format!("cluster{}.leader.perf", idx))
+                        let tracing_options = TracingOptions::builder()
+                            .perf_raw_outfile(format!("cluster{}.perf.data", idx))
+                            .dtrace_outfile(format!("cluster{}.leader.stacks", idx))
                             .fold_outfile(format!("cluster{}.data.folded", idx))
                             .flamegraph_outfile(format!("cluster{}.svg", idx))
-                            .frequency(5)
+                            .frequency(997)
```
Comment (shadaj) [inline hydroflow_plus_test/examples/perf_compute_pi.rs:86]: where did this number come from?

**Review by shadaj — APPROVED** [review]
Hooray!!!

File: hydroflow_plus_test/examples/perf_compute_pi.rs (hunk 3 above)
Comment (MingweiSamuel) [reply hydroflow_plus_test/examples/perf_compute_pi.rs:86]: Was trying to figure out dtrace's weird dsl/syntax
`sudo dtrace -n 'profile-997 /execname == "mybinaryname"/ { @[ustack(100)] = count(); }' -o out.user_stacks`
turns out you can put any number in the 997

Ill change it back

**Pull request merged**